ledger/ledger-utxo/src/main/kotlin/net/corda/v5/ledger/utxo/Contracts.kt

```diff
@@ -26,3 +27,29 @@ interface Contract {
  */
 @CordaSerializable
 interface Command
+
+/**
+ * Defines a verifiable command; that is, a command that implements its own contract verification logic.
+ */
+interface VerifiableCommand : Command {
+
+    /**
+     * Verifies the specified transaction associated with the current contract command.
+     *
+     * @param transaction The transaction to verify.
+     * @param signatories The signatories associated with the current contract command.
+     */
+    fun verify(transaction: UtxoLedgerTransaction, signatories: Iterable<PublicKey>)
+}
+
+/**
+ * Defines a command, and the signatories associated with the specified command.
+ *
+ * @property command The command to verify.
+ * @property signatories The signatory signing keys associated with the specified command.
+ */
+@CordaSerializable
+interface CommandAndSignatories<out T : Command> {
```

Comment:
This is problematic in terms of serialization and indeed doesn't work in C4 due to need to see the required signers before knowing how to deserialize commands. So probably the information structure needs some thinking. I am fine with this as a synthesized view onto the wire/ledger transaction commands, but really they will end up in two (same sized) component groups

Reply:
If we end up with 2 distinct component groups, 1 for commands and 1 for signatories, how will we distinguish which keys belong to which command, or is this to be decided?

Reply:
The component group will be some holder of a list of PublicKeys so one item for one command

Reply:
Main thing is to separate things that need to already know the CPI and make a sandbox to be deserialized and things which can be interrogated without. This is especially important when we do partial reveal data structures.